Evaluate polynomials at equally spaced grid points

postvakje opened the pull request:
Implement Nuttall's algorithm for efficiently evaluating polynomials at
equally spaced grid points.

postvakje committed:
Evaluate polynomials at equally spaced grid points

Implement Nuttall's algorithm for efficiently evaluating polynomials at
equally spaced grid points.

postvakje committed:
Added blank line to end of docstrings of dispersion and dispersonset

Travis CI build failed on these.

postvakje committed:
removed trailing white spaces

smichr commented on sympy/polys/polytools.py:
```diff
@@ -3991,6 +3993,71 @@ def ne(f, g, strict=False):
+
+        Examples
+        ========
+        >>> from sympy import Poly
```
insert a blank line after the `====`

postvakje committed:
added blank lines after ===

smichr commented:
Did you compare times with this method vs simply generating the expressions with subs?

postvakje commented:
Here is a quick test:

from polytools import Poly
from sympy.abc import x
f = Poly(5*x**5+4*x**4+3*x**3+2*x**2+x+1)
For the code:
f.eval_grid(1,1,10000,x)
timeit(10) gives 2.242429741534579
For the code:
[f.subs(x,y) for y in range(1,10001,1)]
timeit(10) gives 8.017799153429223

skirpichev commented:
> For the code

Why do such insane things?  You could do [f(y) for y in range(1,10001,1)] instead, I don't see too much difference:
```python
In [2]: %timeit f.eval_grid(1,1,10000,x)
1 loops, best of 3: 237 ms per loop
In [3]: %timeit [f(y) for y in xrange(1,10001,1)]
1 loops, best of 3: 316 ms per loop
```

postvakje commented:
Yes, I am also getting about 20-30% speedup in integer arithmetic.  Not great, but still significant.  The speedup in floating point arithmetic is much larger, but Nuttall's algorithm computes the next term based on the previous term, so errors can accumulate.  
Also, I found that [f.subs(x,y) for y in arange(1.0,1001.0,1.0)] runs much faster than [f(y) for y in arange(1.0,1001.0,1.0)].  I wonder why that is?

skirpichev commented:
> Yes, I am also getting about 20-30% speedup in integer arithmetic.

I don't see your tests (except for above nonsense), sorry.  And I'm not sure that my tests provide such a proof.

> The speedup in floating point arithmetic is much larger

How, why?

postvakje commented:
Some more tests with integer arithmetic
```
In [19]: f = Poly(5*x**5+4*x**4+3*x**3+2*x**2+x+1)
In [20]: %timeit f.eval_grid(1,1,1000000,x)
1 loops, best of 3: 22.4 s per loop
In [21]: %timeit [f(y) for y in range(1,1000001,1)]
1 loops, best of 3: 31.9 s per loop
```

tests with floating point:
```
In [37]: f2 = Poly(2.2*x**2+4.3*x+2.2)
In [38]: %timeit f2.eval_grid(1.0,1.0,10000,x)
1 loops, best of 3: 214 ms per loop
In [39]: %timeit [f2(y) for y in arange(1.0,10001.0,1.0)]
1 loops, best of 3: 4.96 s per loop
In [40]: %timeit [f2.subs(x,y) for y in arange(1.0,10001.0,1.0)]
1 loops, best of 3: 925 ms per loop
In [41]: %timeit [f2(1.0+y*1.0) for y in range(10000)]
1 loops, best of 3: 5.05 s per loop
In [42]: %timeit [f2.subs(x,1.0+y*1.0) for y in range(10000)]
1 loops, best of 3: 835 ms per loop
```

Note that in this case f2(y) is significantly slower than f2.subs(x,y)

smichr commented:
I'm not convinced that we really need this; I am -0.

postvakje committed:
Merge branch 'master' into eval_grid

Fixed conflicts in:
	sympy/polys/polytools.py
	sympy/polys/tests/test_polytools.py

postvakje committed:
use itertools.accumulate

* Copied itertools.accumulate code from docs.python.org to
sympy.core.compatibility.
* Added note about accumulated errors in floating point arithmetic.
* Timing in python 3.3: about 35% speedup for integer arithmetic and
about 95% speedup for floating point arithmetic.

```python
In [12]: f = Poly(5*x**5+4*x**4+3*x**3+2*x**2+x+1)

In [13]: %timeit f.eval_grid(1,1,1000000,x)
1 loops, best of 3: 18.9 s per loop

In [14]: %timeit [f(y) for y in range(1,1000001,1)]
1 loops, best of 3: 29.9 s per loop

In [15]: f2 = Poly(2.2*x**2+4.3*x+2.2)

In [16]: %timeit f2.eval_grid(1.0,1.0,10000,x)
1 loops, best of 3: 225 ms per loop

In [17]: %timeit [f2(1.0+y*1.0) for y in range(10000)]
1 loops, best of 3: 4.34 s per loop
```

czgdp1807 commented:
Closing this in reference to above comments. Please feel free to re-open this if needed. Thanks for your contributions.
